DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/SecureTokenGenerator.java

```diff
@@ -0,0 +1,137 @@
+    // Letter "I" has been removed, as it shows up too similar to letter "l" on bridge
+    private static final String UPPERCASE_ALPHA = "ABCDEFGHJKLMNPQRSTUVWXYZ";
+    // Bridge password must contain at least one lowercase letter (a-z)
+    // Letter "l" has been removed, as it shows up too similar to letter "I" on bridge
+    private static final String LOWERCASE_ALPHA = "abcdefghijkmnopqrstuvwxyz";
+    // "0" has been removed, as it shows up too similar to letter "O" on bridge
+    private static final String NUMERIC = "123456789";
+    // Bridge password must contain at least one symbol ( !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ );
+    // This subset of symbols was chosen because they would be easier to communicate over the phone
+    private static final String SYMBOL = "!#$%&'*+,-.:;=?@_";
+
+    private static final String ALPHANUMERIC = UPPERCASE_ALPHA + LOWERCASE_ALPHA + NUMERIC;
+    // ALPHANUMERIC was added 3 times to decrease the number of symbols in a password
+    private static final String PASSWORD = ALPHANUMERIC + ALPHANUMERIC + ALPHANUMERIC + SYMBOL;
+
+    public static final SecureTokenGenerator INSTANCE = new SecureTokenGenerator();
+
+    public static final SecureTokenGenerator ARC_ID_INSTANCE =
+            new SecureTokenGenerator(6, new SecureRandom(), NUMERIC);
+
+    /**
+     * I used this website https://asecuritysite.com/encryption/passes and using our parameters,
+     * it said a 9 character password has 66,540,410,775,079,424 available passwords,
+     * and it would take 2108.59 years to crack if requests were sent.
+     */
+    public static final SecureTokenGenerator BRIDGE_PASSWORD =
+            new SecureTokenGenerator(9, new SecureRandom(), PASSWORD);
+
+    private final Random random;
+    private final char[] characters;
+    public int getTokenLength() {
+        return characters.length;
+    }
+    private final int length;
+
+    private SecureTokenGenerator(int length, Random random, String characters) {
+        checkArgument(length > 1);
+        checkArgument(characters != null && characters.length() >= 2);
+
+        this.length = length;
+        this.random = Objects.requireNonNull(random);
+        this.characters = characters.toCharArray();
+    }
+
+    /**
+     * Create session identifiers. This is 4.36e+37 unique values,
+     * which is enough for a good sessions key.
+     */
+    private SecureTokenGenerator() {
+        this(21, new SecureRandom(), ALPHANUMERIC);
+    }
+
+    public String nextToken() {
+        final char[] buffer = new char[length];
+        for (int i = 0; i < buffer.length; ++i) {
+            buffer[i] = characters[random.nextInt(characters.length)];
+        }
+        return new String(buffer);
+    }
+
+    /**
+     * Guaranteed to generate a random password string compatible with Bridge
+     * @return a random password string compatible with Bridge
+     */
+    public String nextBridgePassword() {
```

Comment:
Nice! Thank you. 

First, I love that you did those calculations @brucehoff a true computer scientist!

Also, I switched over to Alx's PasswordGenerator class.  It is an order of magnitude more efficient than mine, especially when run at 1,000,000 iterations.  Doing a second pass using random indices of the password is a smart way to guarantee it meets the criteria.

One thing I did notice however was that the PasswordGenerator's passwords only ever have exactly one SYMBOLIC character. I wanted to double check that should be fine for DIAN? @alxdarksage  I could switch it to include the SYMBOLIC characters on the first pass; however, that could lead to upwards of 6 symbolic characters in my 9 character password, and that seems like bad user experience to type in.

Reply:
I think our generator uses all symbols in the first pass and then ensures there's at least one, but because you have users typing these in, it is probably fine to only have one of the special characters. Again, this is still vastly more secure than what we were going to do initially!